File: testing/src/simple/Ellipse.java
package simple;

import remixlab.input.*;
import remixlab.input.event.*;
import processing.core.*;

/**
 * Created by pierre on 12/22/16.
 */
public class Ellipse extends GrabberObject {
  PApplet parent;
  public float radiusX, radiusY;
  public PVector center;
  public int colour;
  public int contourColour;
  public int sWeight;

  public Ellipse(PApplet p, Agent agent) {
    super(agent);
    parent = p;
    agent.addGrabber(this);
    setColor();
    setPosition();
    sWeight = 4;
    contourColour = parent.color(0, 0, 0);
  }

  public Ellipse(Agent agent, PVector c, float r) {
    super(agent);
    radiusX = r;
    radiusY = r;
    center = c;
    setColor();
    sWeight = 4;
  }

  public void setColor() {
    setColor(parent.color(parent.random(0, 255), parent.random(0, 255), parent.random(0, 255), parent.random(100, 200)));
  }

  public void setColor(int myC) {
    colour = myC;
  }

  public void setPosition(float x, float y) {
    setPositionAndRadii(new PVector(x, y), radiusX, radiusY);
  }

  public void setPositionAndRadii(PVector p, float rx, float ry) {
    center = p;
    radiusX = rx;
    radiusY = ry;
  }

  public void setPosition() {
    float maxRadius = 50;
    float low = maxRadius;
    float highX = 800 - maxRadius;
    float highY = 800 - maxRadius;
    float r = parent.random(20, maxRadius);
    setPositionAndRadii(new PVector(parent.random(low, highX), parent.random(low, highY)), r, r);
  }

  public void draw() {
    draw(colour);
  }

  public void draw(int c) {
    parent.pushStyle();
    parent.stroke(contourColour);
    parent.strokeWeight(sWeight);
    parent.fill(c);
    parent.ellipse(center.x, center.y, 2*radiusX, 2*radiusY);
    parent.popStyle();
  }

  @Override
  public boolean track(MotionEvent2 event) {
    return track(event.x(), event.y());
  }

  @Override
  public boolean track(TapEvent event) {
    return track(event.x(), event.y());
  }

  public boolean track(float x, float y) {
    return(parent.pow((x - center.x), 2)/parent.pow(radiusX, 2) + parent.pow((y - center.y), 2)/parent.pow(radiusY, 2) <= 1);
  }

  @Override
  public void interact(TapEvent event) {
    if ( event.id() == PApplet.LEFT )
      setColor();
    else if ( event.id() == PApplet.RIGHT ) {
      if (sWeight > 1)
        sWeight--;
    } else
      sWeight++;
  }

  @Override
  public void interact(MotionEvent2 event) {
    if ( event.id() == PApplet.LEFT )
      setPosition();
    else if ( event.id() == PApplet.RIGHT ) {
      radiusX += event.dx();
      radiusY += event.dy();
    }
  }
}

File: examples/ShortcutBasedEventParsing/Ellipse.js
class Ellipse {

  constructor(center = null, radius = null) {

    if ((radius === null && center === null)) {
      this.radiusX, this.radiusY, this.center;
      this.setPosition();

    } else {
      this.radiusX = radius;
      this.radiusY = radius;
      this.center = center;
    }
    this.colour;
    this.contourColour = color(0, 0, 0);
    this.move = false;
    this.sWeight = 4;
    this.setColor();
    this.setMouseDragBindings();
  }

  setColor(myC = null) {
    if(myC !== null)
      this.colour = myC;
    else
      this.colour = color(random(0, 255), random(0, 255), random(0, 255), random(100, 200));
  }

  setPosition(event = null) {
    if(event === null){
      const maxRadius = 50;
      const low = maxRadius;
      const highX = width - maxRadius;
      const highY = height - maxRadius;
      const r = random(20, maxRadius);
      this.setPositionAndRadii(new p5.Vector(random(low, highX), random(low, highY)), r, r);
    }else {
      this.setPositionAndRadii(new p5.Vector(event.x(), event.y()), this.radiusX, this.radiusY);
    }
  }

  setShape(event, dof1 = true) {
      this.radiusX += event.dx();
      this.radiusY += dof1 ? event.dx() : event.dy();
  }

  setPositionAndRadii(p, rx, ry) {
    this.center = p;
    this.radiusX = floor(rx);
    this.radiusY = floor(ry);
  }


  draw(c = this.colour) {
    push();
    stroke(this.contourColour);
    strokeWeight(this.sWeight);
    fill(c);
    ellipse(this.center.x, this.center.y, 2*this.radiusX, 2*this.radiusY);
    pop();
  }

  setMouseDragBindings() {
    this.move = false;
  }

  setMouseMoveBindings() {
    this.move = true;
  }

  interact(event) {
    if (event.shortcut().matches(new bias.Shortcut({ id: MouseAgent.WHEEL, modifiers: bias.Event.CTRL }))) {
      this.setShape(event);
    } else if (event.shortcut().matches(new bias.event.TapShortcut({ id: MouseAgent.CLICK, count: 1 }))) {
      this.setColor();
    } else {
      this.motion2Interaction(event);
    }
  }

  track(event) {
    //we're sure it must be a Tap or a MotionEvent
    const x = event.x();
    const y = event.y()
    return (pow((x - this.center.x), 2)/pow(this.radiusX, 2) + pow((y - this.center.y), 2)/pow(this.radiusY, 2) <= 1);
  }

  motion2Interaction(event) {
    if (this.move) {
      if (event.shortcut().matches(new bias.Shortcut({ id: MouseAgent.NO_BUTTON }))) {
        this.setPosition(event);
      }
    } else {
      if (event.shortcut().matches(new bias.Shortcut({ id: MouseAgent.LEFT })))
        this.setPosition(event);
    }
    if (event.shortcut().matches(new bias.Shortcut({ id: MouseAgent.RIGHT }))) {
      this.setShape(event, false);
    }
  }
}